pkg/migration/migrations/raft_applied_index_term.go

```diff
@@ -51,9 +69,32 @@ func raftAppliedIndexTermMigration(
 			if bytes.Compare(desc.StartKey, keys.LocalMax) < 0 {
 				start, _ = keys.Addr(keys.LocalMax)
 			}
-			if err := deps.DB.Migrate(ctx, start, end, cv.Version); err != nil {
+
+			// Skip any ranges below the watermark.
+			if bytes.Compare(end, watermark) <= 0 {
+				continue
+			}
+
+			// Migrate the range, with a few retries.
+			if err := retryOpts.Do(ctx, func(ctx context.Context) error {
+				err := deps.DB.Migrate(ctx, start, end, cv.Version)
+				if err != nil {
+					log.Errorf(ctx, "range %d migration failed, retrying: %s", desc.RangeID, err)
+				}
+				return err
+			}); err != nil {
 				return err
 			}
+
+			// Raise the watermark.
+			progress.Watermark = end
+		}
+
+		// Persist the migration's progress.
+		if len(progress.Watermark) > 0 {
+			if err := job.SetProgress(ctx, nil, progress); err != nil {
```

Comment:
Isn't this a row write to the jobs table? Would we be writing once per batch of range descriptors? That might be expensive for clusters with a large number of ranges and possible accrue a ton of MVCC garbage.

Reply:
It is. But with a page size of 200, that means that a cluster with a million ranges will only see 5000 writes. In any case, I've changed it to only persist every 5 batches, so it's now down to 1000 writes.